### PR TITLE
SmartQuerySetMixin - Add ordering by  if not present.

### PR DESCRIPTION
**Pull request by vichi99**

### CODE REVIEW

File: chamber/models/base.py [hunk 1]
```diff
@@ -52,6 +52,28 @@ def last(self, *field_names):
         else:
             return super().last()
 
+    def order_by(self, *field_names):
```
Comment (vichi99) [inline chamber/models/base.py:56]: Pripadne muzeme pridat default atribut do tridy, kterym to muzeme vypnout.

File: chamber/models/base.py [hunk 2]
```diff
@@ -52,6 +52,28 @@ def last(self, *field_names):
...
+        """
+        If field_names is not empty, it adds field pk to it.
+        Adding pk reduces the error when ordering by the same values.
```
Comment (rubickcz) [inline chamber/models/base.py]: `Adding pk makes ordering deterministic, when column has multiple same values.`

File: chamber/version.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-VERSION = (0, 6, 17, 1)
+VERSION = (0, 6, 17, 2)
```
Comment (rubickcz) [inline chamber/version.py]: I believe this is done automatically during release.

File: chamber/models/base.py (hunk 1 above)
Comment (rubickcz) [reply chamber/models/base.py:56]: Pls use english language.

File: chamber/models/base.py [hunk 4]
```diff
@@ -52,6 +52,28 @@ def last(self, *field_names):
...
+            and not [
+                field for field in field_names if
+                (isinstance(field, str) and field in ['pk', '-pk'])
+                or
+                (hasattr(field, 'expression') and field.expression.name == 'pk')
+            ]
+        ):
+            if hasattr(field_names[-1], 'descending'):
+                # Without this, the column could not be sorted in both directions (asc,dsc) if it had
+                # the same values everywhere and the user sorted them according to OrderBy.
```
Comment (rubickcz) [inline chamber/models/base.py]: This comment is redundant. All information needed is in docstring already.

File: chamber/models/base.py [hunk 5]
```diff
@@ -52,6 +52,28 @@ def last(self, *field_names):
...
+        if (
+            field_names
+            and not [
+                field for field in field_names if
+                (isinstance(field, str) and field in ['pk', '-pk'])
+                or
+                (hasattr(field, 'expression') and field.expression.name == 'pk')
+            ]
+        ):
+            if hasattr(field_names[-1], 'descending'):
```
Comment (rubickcz) [inline chamber/models/base.py]: What if last field name is string? In that case, you are ignoring the order.

File: chamber/models/base.py (hunk 1 above)
Comment (rubickcz) [reply chamber/models/base.py:56]: Also please add tests.

File: chamber/models/base.py (hunk 5 above)
Comment (vichi99) [reply chamber/models/base.py]: Yes. When the last element is OrderBy, it is added by the user, here it is necessary to swap asc/desc. In the case of strings, it is a programmer-set stamping, here it is not necessary.

File: chamber/models/base.py (hunk 1 above)
Comment (vichi99) [reply chamber/models/base.py:56]: Tests added.

File: README.md [hunk 6]
```diff
@@ -35,6 +35,7 @@ You are welcomed to contribute at https://github.com/druids/django-chamber. Ther
 
 * documentation of this library is a work in progress, needs a lot of attention
 * SmartModel extends the AuditModel and therefore always adds `created_at` and `changed_at` fields to the model which is not always desirable
+* If occur problem during `make install` like `Unable to find magic library` try install libmagic  `brew install libmagic` - [https://filemagic.readthedocs.io/en/latest/guide.html#installation](https://filemagic.readthedocs.io/en/latest/guide.html#installation)
```
Comment (rubickcz) [inline README.md]: - problem occurs
- try installing

File: chamber/models/base.py [hunk 7]
```diff
@@ -52,6 +53,30 @@ def last(self, *field_names):
         else:
             return super().last()
 
+    def order_by(self, *field_names):
+        """
+        If field_names do not contain unique fields (pk, id, created_at,  changed_at), '-' + pk is added.
+        Adding pk (by model) makes ordering deterministic, when column has multiple same values.
+        """
+        pk = self.model._meta.pk.name if self.model else 'pk'
```
Comment (rubickcz) [inline chamber/models/base.py:61]: Does it return `id`?

File: chamber/models/base.py [hunk 8]
```diff
@@ -52,6 +53,30 @@ def last(self, *field_names):
...
+        unique_fields = [pk, 'pk', 'id',  'created_at',  'changed_at']
```
Comment (rubickcz) [inline chamber/models/base.py]: `created_at` and `changed_at` are not unique fields. The line should be just:
```
unique_fields = {pk, 'pk'}
```

File: chamber/models/base.py [hunk 9]
```diff
@@ -52,6 +53,30 @@ def last(self, *field_names):
...
+        if (
+            not [
+                field for field in field_names or tuple() if
```
Comment (rubickcz) [inline chamber/models/base.py]: `or tuple()` is not needed

File: chamber/models/base.py [hunk 10]
```diff
@@ -52,6 +53,30 @@ def last(self, *field_names):
...
+                (isinstance(field, str) and field.replace('-', '') in unique_fields)
+                or
+                (hasattr(field, 'expression') and field.expression.name.replace('-', '') in unique_fields)
+            ]
+        ):
+            field_names += ('-' + pk,)
```
Comment (rubickcz) [inline chamber/models/base.py]: Just add `pk`, no need to do reverse order.

File: chamber/models/base.py [hunk 11]
```diff
@@ -52,6 +53,30 @@ def last(self, *field_names):
...
+        return super().order_by(*field_names)
+
+    def order_by_without_pk(self, *field_names):
+        """
+        Default Django order_by method without automatic adding the pk field.
```
Comment (rubickcz) [inline chamber/models/base.py]: adding **of** the pk field

File: chamber/models/base.py (hunk 7 above)
Comment (vichi99) [reply chamber/models/base.py:61]: Yes by default.

File: chamber/models/base.py (hunk 5 above)
Comment (vichi99) [reply chamber/models/base.py]: is no longer relevant

File: chamber/models/base.py (hunk 4 above)
Comment (vichi99) [reply chamber/models/base.py]: is no longer relevant

**Comment by vichi99** [comment]
decided not to solve the problem or to go with another solution https://github.com/skip-pay/django-pyston/pull/5

**Pull request closed without merging**